### PR TITLE
Provides a connector for the Repository and IDB within a Jupiter instance.

### PR DESCRIPTION
**Pull request by andyHa**

### CODE REVIEW

File: src/main/java/sirius/biz/jupiter/JupiterSync.java [hunk 1]
```diff
@@ -29,12 +38,21 @@
...
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
 
 /**
  * In charge of updating the configuration and the repository of attached Jupiter instances.
+ * <p>
+ * Next to iterating over all Jupiter instances and computer their config via {@link JupiterConfigUpdater}, this also
```
Comment (sabieber) [inline src/main/java/sirius/biz/jupiter/JupiterSync.java:56]: computer -> computing

File: src/main/java/sirius/biz/jupiter/JupiterSync.java [hunk 2]
```diff
@@ -164,4 +215,249 @@ private String asYaml(Map<String, Object> config) {
...
+                                            .withSystemErrorMessage(
+                                                    "Failed to check the local repository %s for %s: %s (%s)",
+                                                    localRepoSpaceName,
+                                                    connection.getName())
+                                            .handle());
+        }
+    }
+
+    private void visitLocalDirectory(ProcessContext processContext,
+                                     String prefix,
```
Comment (sabieber) [inline src/main/java/sirius/biz/jupiter/JupiterSync.java:447]: make this nullable? So we dont have to pass "" above

File: src/test/resources/jupiter-test/repository/countries.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  de: "Deutschland"
+  en: "Germany"
+---
+code: "A"
+iso:
+  two: "at"
+  three: "aut"
+name:
+  de: "Österreich"
+  en: "Austria"
```
Comment (sabieber) [inline src/test/resources/jupiter-test/repository/countries.yml]: no empty line at end of file

File: src/main/java/sirius/biz/jupiter/IDBTable.java [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+     * @param jupiter the jupiter connection to wrap
+     * @param name    the name of the table to wrap
+     */
+    public IDBTable(JupiterConnector jupiter, String name) {
+        this.jupiter = jupiter;
+        this.name = name;
+    }
+
+    /**
+     * Used as fluent tool to build queries.
```
Comment (sabieber) [inline src/main/java/sirius/biz/jupiter/IDBTable.java:58]: Used -> Uses

File: src/main/java/sirius/biz/jupiter/IDBTable.java [hunk 5]
```diff
@@ -0,0 +1,563 @@
...
+
+        /**
+         * Specifies the paths to look for a given value.
+         * <p>
+         * Note that this will perform an exact search (the search value has to match the value in the field).
+         *
+         * @param paths one or more paths to search in
+         * @return the builder itself for fluent method calls
+         */
+        public QueryBuilder lookupPaths(String... paths) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/jupiter/IDBTable.java:76]: To ensure exactly one of the three modes is used at the same time these could be declared as static functions that are the only ways to get an instance of the builder.

File: src/main/java/sirius/biz/jupiter/JupiterConnector.java [hunk 6]
```diff
@@ -30,6 +30,8 @@
  * <p>
  * If a failover is performed, we will continue to use the fallback for up to 60s before we attempt to switch
  * back to the main instance.
+ * <p>
+ * A connector is usually obtained via {@link Jupiter#getConnector(String)} ot {@link Jupiter#getDefault()}.
```
Comment (sabieber) [inline src/main/java/sirius/biz/jupiter/JupiterConnector.java:34]: ot -> or

File: src/main/java/sirius/biz/jupiter/JupiterDataProvider.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ * Copyright by scireum GmbH
+ * http://www.scireum.de - info@scireum.de
+ */
+
+package sirius.biz.jupiter;
+
+import java.io.OutputStream;
+
+/**
+ * Provides data to be stored in a file inthe local Jupiter repository.
```
Comment (sabieber) [inline src/main/java/sirius/biz/jupiter/JupiterDataProvider.java:14]: inthe -> in the

File: src/main/java/sirius/biz/jupiter/JupiterDataProvider.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ * <p>
+ * The {@link JupiterSync} will invoke all data providers each night to update their file contents in the local
+ * repository. This will then be synchronized with the repository in the Jupiter instance itself so that
+ * InfoGraphDB and other consumers can load this data.
+ */
+public interface JupiterDataProvider {
```
Comment (sabieber) [inline src/main/java/sirius/biz/jupiter/JupiterDataProvider.java:20]: AutoRegister?

File: src/main/java/sirius/biz/jupiter/JupiterSync.java [hunk 9]
```diff
@@ -494,6 +503,34 @@ private void handleLocalFile(ProcessContext processContext,
         filesToDelete.remove(effectiveFileName);
     }
 
+    private void awaitNextEpoch(ProcessContext processContext, JupiterConnector connector) {
+        connector.repository().requestEpoch();
+        int retries = 30;
```
Comment (sabieber) [inline src/main/java/sirius/biz/jupiter/JupiterSync.java:511]: magic numbers

File: src/main/java/sirius/biz/jupiter/JupiterSync.java [hunk 10]
```diff
@@ -494,6 +503,34 @@ private void handleLocalFile(ProcessContext processContext,
...
+        Monoflop stateUpdate = Monoflop.create();
+        while (processContext.isActive() && retries-- > 0 && !connector.repository().isEpochSync()) {
+            if (stateUpdate.firstCall()) {
+                processContext.setCurrentStateMessage(Strings.apply("Waiting for the repository of %s to be synced...",
+                                                                    connector.getName()));
+            }
+            Wait.seconds(2);
```
Comment (sabieber) [inline src/main/java/sirius/biz/jupiter/JupiterSync.java:518]: magic numbers

File: src/main/java/sirius/biz/jupiter/Repository.java [hunk 11]
```diff
@@ -109,4 +111,35 @@ private RepositoryFile parseMetadataRow(Object obj) {
...
+
+    /**
+     * Determines if the foreground epoch and background epoch of Jupiter are equal.
+     * <p>
+     * If previously {@link #requestEpoch()} has been called, this can be used to guarantee, that the background
+     * actor of Jupiter is idle/has completed all tasks.
+     *
+     * @return <tt>true</tt> if the epochs are equal (the background system is idle), <tt>false</tt> otherwise
+     */
+    public boolean isEpochSync() {
```
Comment (sabieber) [inline src/main/java/sirius/biz/jupiter/Repository.java:138]: isEpochInSync?

File: src/main/java/sirius/biz/jupiter/Jupiter.java [hunk 12]
```diff
@@ -171,15 +179,55 @@ public static Object read(Object obj) {
...
                                 "ms");
     }
+
+    /**
+     * Fetches a locally cached object based on Jupiter data which is relatively small.
+     * <p>
+     * This is a larger cache for (relatively) small objects like strings or tuples.
+     * <p>
+     * Note that the underlying cache is automatically cleared once the Jupiter repository is synced, so that
+     * this cache shoudln't contain any stale data.
```
Comment (sabieber) [inline src/main/java/sirius/biz/jupiter/Jupiter.java:196]: shoudln't -> shouldn't

File: src/main/java/sirius/biz/jupiter/Jupiter.java [hunk 13]
```diff
@@ -171,15 +179,55 @@ public static Object read(Object obj) {
...
+     */
+    @SuppressWarnings("unchecked")
+    public <V> V fetchFromSmallCache(String key, Supplier<V> valueComputer) {
+        return (V) LOCAL_SMALL_CACHE.get(key, ignored -> valueComputer.get());
+    }
+
+    /**
+     * Fetches a locally cached object based on Jupiter data which is relatively large/complex.
+     * <p>
+     * This is a larger cache for (relatively) large/complex compound objects.
```
Comment (sabieber) [inline src/main/java/sirius/biz/jupiter/Jupiter.java:211]: larger -> smaller?

**Review by sabieber — CHANGES_REQUESTED** [review]

File: src/main/java/sirius/biz/jupiter/JupiterSync.java (hunk 2 above)
Comment (andyHa) [reply src/main/java/sirius/biz/jupiter/JupiterSync.java:447]: nope null + "/" != "" + "/"

File: src/main/java/sirius/biz/jupiter/IDBTable.java (hunk 5 above)
Comment (andyHa) [reply src/main/java/sirius/biz/jupiter/IDBTable.java:76]: no you can skip searching entirely and retrieve all rows...

File: src/main/java/sirius/biz/jupiter/JupiterSync.java (hunk 2 above)
Comment (sabieber) [reply src/main/java/sirius/biz/jupiter/JupiterSync.java:468]: Yes my intention was to suggest handling a passed null inside the method with a simple if or ternary

File: src/main/java/sirius/biz/jupiter/IDBTable.java (hunk 5 above)
Comment (sabieber) [reply src/main/java/sirius/biz/jupiter/IDBTable.java:76]: okay +1

**Review by jmuscireum — APPROVED** [review]

File: src/main/java/sirius/biz/jupiter/JupiterConnector.java (hunk 6 above)
Comment (andyHa) [inline src/main/java/sirius/biz/jupiter/JupiterConnector.java:34]:
```suggestion
 * A connector is usually obtained via {@link Jupiter#getConnector(String)} or {@link Jupiter#getDefault()}.
```

File: src/main/java/sirius/biz/jupiter/JupiterDataProvider.java (hunk 7 above)
Comment (andyHa) [inline src/main/java/sirius/biz/jupiter/JupiterDataProvider.java:14]:
```suggestion
 * Provides data to be stored in a file in the local Jupiter repository.
```

File: src/main/java/sirius/biz/jupiter/JupiterDataProvider.java (hunk 8 above)
Comment (andyHa) [reply src/main/java/sirius/biz/jupiter/JupiterDataProvider.java:20]: Interfaces are "inherently autoregister" - as long as we don't plan to use a common abstract base class, this isn't strictly required - on the other hand, it doesn't hurt

Comment (andyHa) [reply src/main/java/sirius/biz/jupiter/JupiterDataProvider.java:20]: ...and it also signals that it is intended to be registered...

File: src/main/java/sirius/biz/jupiter/JupiterSync.java [hunk 14]
```diff
@@ -76,39 +146,102 @@ public String getName() {
...
-            tasks.defaultExecutor().start(this::performSync);
-        }
+        runInStandbyProcess(processContext -> performSyncInProcess(processContext, true, true, true));
     }
 
     @Override
     public void started() {
         if (automaticUpdate) {
-            elastic.getReadyFuture().onSuccess(this::performSync);
+            // As re run in the "readiness callback" of elastic, we rather fork a thread here, as synchronizing the
```
Comment (bwiedmann) [inline src/main/java/sirius/biz/jupiter/JupiterSync.java:156]:
```suggestion
            // As we run in the "readiness callback" of elastic, we rather fork a thread here, as synchronizing the
```

**Review by bwiedmann — CHANGES_REQUESTED** [review]

**Review by sabieber — APPROVED** [review]

**Review by idlira — APPROVED** [review]

File: src/test/resources/docker-compose.yml [hunk 15]
```diff
@@ -7,8 +7,14 @@ services:
       - "6379"
     hostname: redis
 
+  redis-jupiter:
+    image: scireum/jupiter-io:0.2.1
```
Comment (mkeckmkeck) [inline src/test/resources/docker-compose.yml:11]:
```suggestion
    image: scireum/jupiter-io:0.3.0
```

**Review by mkeckmkeck — CHANGES_REQUESTED** [review]

**Review by mkeckmkeck — APPROVED** [review]

**Review by sabieber — APPROVED** [review]

**Review by bwiedmann — APPROVED** [review]

**Review by mkeckmkeck — APPROVED** [review]

**Pull request merged**